### ISSUE TITLE
DataTable option that allows the state of the table to last between refreshes / re-rendering

### ISSUE THREAD

**Issue by XaeroDegreaz** [TheCoder4eu/BootsFaces-OSP#279]
I thought of this while working with a DataTable that had a column with buttons in it that allowed me to re-order the items in the backing list, then re-render the table via AJAX. The problem comes when you're re-ordering items on pages other than the 1st page. When you re-render the DataTable, you start back on page one.

It would be nice to have an attribute for the DataTable that allows you to bind the state of of the component (page, search input, currently sorted column, results per page dropdown, etc) to a backing bean. This would allow the table properties to inherit the scope of the backing bean (`@SessionScoped`, `@ViewScoped`, etc).

I haven't looked through the actual BootsFaces code for this component but here's how I imagine it would work:

``` java
@ViewScoped
@Named
public class MyBackingBean implements Serializable {
    private List<SomeObject> someObjectList;
    private Map<String, Object> dataTableProperties;

    @PostConstruct
    void postConstruct() {
        someObjectList = SomeService.gimmeAList();
    }    

    public List<String> getSomeObjectList() {
        return someObjectList;
    }    

    public Map<String, Object> getDataTableProperties(){
        return dataTableProperties;
    }

    public void setDataTableProperties(Map<String, Object> dataTableProperties){
        this.dataTableProperties = dataTableProperties;
    }
}
```

``` xml
<b:dataTable
    value="#{myBackingBean.someObjectList}"
    var="obj"
    dataTableProperties="#{myBackingBean.dataTableProperties}>
    ....
    ....
</b:dataTable>
```

If the attribute is not null in the DataTable component, then it would read / write to that map for all of it's properties :)

Hope I was clear.

**Comment by XaeroDegreaz** [comment]
Ahh ok I see -- the actual DataTable state is maintained client-side (which is why it's nice and fast). Looks like you would need to hack on their library in order for it to send state back to backing bean, and have it populate correctly when it renders again.

**Comment by XaeroDegreaz** [comment]
Ok, so no hacking on their code required. According to their [API](https://datatables.net/reference/api/page()), when you initialize the DataTable, you can also interact some more with their API.

So if you were to use my example above you could modify [this class](https://github.com/TheCoder4eu/BootsFaces-OSP/blob/master/src/main/java/net/bootsfaces/component/dataTable/DataTableRenderer.java#L175-L183)

and have something like:

```java
DataTable dataTable = (DataTable) component;
Map<String, Object> dataTableProperties = 
    (Map<String, Object>)dataTable.getAttributes().get("dataTableProperties");
//# null check
int page = dataTableProperties.get("currentPage");
/////
ResponseWriter rw = context.getResponseWriter();
String clientId = dataTable.getClientId().replace(":", "");
rw.endElement("table");
Tooltip.activateTooltips(context, dataTable);
rw.startElement("script", component);
rw.writeText("$(document).ready(function() {$('." + clientId + "Table" + "').DataTable().page("+page+");} );",
		null);
rw.endElement("script");
```

Now we gotta figure out how to modify the page buttons on the DataTable JS to add some onclick BSF AJAX calls, or something, to talk to the backing bean marking the page. 

Makes sense what I'm getting at so far? Exciting :D

**Comment by XaeroDegreaz** [comment]
I have a partial working concept in a fork.

https://github.com/TheCoder4eu/BootsFaces-OSP/compare/master...XaeroDegreaz:datatableproperties-279

Next step is to get the [event listener for paging](http://datatables.net/reference/event/page) to send some sort of BSF based AJAX call to either the DataTable instance, or the backing bean. I think it would be more proper to have it send to the actual DataTable instance instead of needing to create an additional listener on the backing bean.

Who knows, you guys are the experts :D

**Comment by stephanrauh** [comment]
Sounds good! I prefer sending the AJAX request to the DataTable instance.

You can send an AJAX request to the server by calling `BSF.ajax()` in your JavaScript code. I suggest you simple take the JS code from a `b:commandButton` and adapt it. On the server side, your can add some code to `DataTableRenderer::decode` to fetch the AJAX request and extract the page number. More precisely, you have to override the method `AJAXRenderer::decode`. I've done something similar for `b:tabView` and `b:selectBooleanCheckbox`.

**Comment by stephanrauh** [comment]
BTW, chances are we're going to use the Chinese dataTable in BootsFaces 0.9+. But I believe it works similar, so your work is going to be helpful even if we're going to replace the current dataTable widget by another client-side implementation.

**Comment by XaeroDegreaz** [comment]
Ok, thanks for the helpful info -- I look forward to start in on this again after I get off work :)

What do you mean by the "Chinese" datatable?

**Comment by XaeroDegreaz** [comment]
@stephanrauh - I have it working! It loads and saves state to the backing bean now! Haha I feel like a champ.

**Backing bean:**
```java
package com.dobydigital;

import net.bootsfaces.component.dataTable.DataTable.DataTablePropertyType;
import javax.annotation.PostConstruct;
import javax.enterprise.context.SessionScoped;
import javax.inject.Named;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by XaeroDegreaz on 1/29/2016.
 */
@Named
@SessionScoped
public class DataTableTest implements Serializable
{
    private List<TestObject> objects;
    private Map<DataTablePropertyType, Object> dataTableProperties;

    @PostConstruct
    void postConstruct()
    {
        objects = new ArrayList<>( 100 );
        for ( int i = 0; i < 100; i++ )
        {
            objects.add( new TestObject( "Object " + i ) );
        }
        dataTableProperties = new HashMap<>();
    }

    public List<TestObject> getObjects()
    {
        return objects;
    }

    public void setObjects( List<TestObject> objects )
    {
        this.objects = objects;
    }

    public Map<DataTablePropertyType, Object> getDataTableProperties()
    {
        return dataTableProperties;
    }

    public void setDataTableProperties( Map<DataTablePropertyType, Object> dataTableProperties )
    {
        this.dataTableProperties = dataTableProperties;
    }
}
```

**Usage**
```xml
<b:dataTable
    value="#{dataTableTest.objects}"
    dataTableProperties="#{dataTableTest.dataTableProperties}"
    var="obj">
    <h:column>
        <f:facet name="header">Name</f:facet>
        #{obj.name}
    </h:column>
</b:dataTable>
```

Diff:

https://github.com/TheCoder4eu/BootsFaces-OSP/compare/master...XaeroDegreaz:datatableproperties-279

Should I just start a pull request so we can have in-line viewing here, or do you wanna keep clicking my links :P

**Comment by stephanrauh** [comment]
Looks good at first glance! Make it a pull request.

BTW, there's a getter to read a single parameter. I guess you can simplify your code a bit. If I've got it right, you always parse the entire list.

Regards,
Stephan

**Comment by XaeroDegreaz** [comment]
There's a getter to read a single parameter? What parameter would you be referring to?

**Comment by XaeroDegreaz** [comment]
Ok, I think it's ready to have a look at.

I'm noticing some strange `log.warn` messages after I manipulate the DataTable without refreshing. I'm not entirely sure of what's happening. It's worth noting that nothing seems to be broken with the table, and all of the functionality (including the new stuff) works just fine.

```
Feb 01, 2016 10:23:17 PM com.sun.faces.application.view.FaceletPartialStateManagementStrategy saveDynamicActions
WARNING: Unable to save dynamic action with clientId 'j_id18' because the UIComponent cannot be found
```

After I flip pages a couple of times, those start spitting out. So far I get like 5 or 6 each time.

The ID numbers seem to increment by 1 each time I flip through pages, change page length, or change the search term.

I've confirmed that it only seems to happen when I have the event listeners in place -- when I comment them out it seems to put log those warnings =\

**Comment by stephanrauh** [comment]
Don't worry about the warning. It's caused by OmniFaces and has been fixed in the current 2.x branch of OmniFaces. Anyways, it does do any harm.

**Comment by XaeroDegreaz** [comment]
O_O you merged this directly into master already. Guess you gotta change the milestone now?

**Comment by stephanrauh** [comment]
Yeah... we decided to use the master as the bug-fix version, and your PR feels like a but fix :).

**Comment by stephanrauh** [comment]
Also merged the PR into the 0.9.0 branch.

**Comment by stephanrauh** [comment]
I just saw this option: https://datatables.net/examples/basic_init/state_save.html

It seems this solves the problem much more simple and elegant. So I'd like to remove your code (which is a pity!) and replace it with a simple client-side option. Any objections?

**Comment by XaeroDegreaz** [comment]
I havent looked at their state-save API but absolutely no objections -- simplest solution is usually the best and better to let the framework take care of it instead of rolling your own :)

**Comment by stephanrauh** [comment]
The issue has been solved by #416 .
